### DISCUSSION TITLE
How to export a detectron2 model into onnx format ?

### DISCUSSION THREAD

**Discussion by ghost** [facebookresearch/detectron2#3608] [Q&A]
# Ask a Question

### Question
I'm trying to export a detectron2 model in onnx format (model.onnx) and do inference with onnxruntime using the exported file (model.onnx).

To do so, i tried to export the famous instance segmentation model provided by detectron2 ( model_zoo ), i succeeded to get the model.onnx file ( i'm not sure if i did correctly though ) but i can't use it to do inference, can anyone tell me what i'm doing wrong ?

Here is a shared google colab notebook containing my code https://colab.research.google.com/drive/11BB-iufQCmwHoRzS4iBFwH7SqasDOik8?usp=sharing

Note 1 : In order to modify the file, you'll have to make a copy of it in your drive !
Note 2 : I raised this issue in the onnx github page, and their answer was : 
```
Hi @mountassir-el,
As the error mentioned, AliasWithName is not an official ONNX op so there is no registered schema for it.
You can check all official ONNX ops here: https://github.com/onnx/onnx/blob/master/docs/Operators.md Thus,
this should be a question for the converter.
Please raise this question in detectron2 to ask the exporter only to produce official ONNX ops. 
Close here now. Please reopen it if you still have other ONNX questions. Thanks!
```

And here is the error message ( you can find it in the notebook but i wanted to add it just for SEO purposes, in case someone is looking for a solution to the same problem ) 

```
Fail
Traceback (most recent call last)

<ipython-input-18-8bd369dd3daf> in <module>()
      1 import onnxruntime as rt
      2 
----> 3 session = rt.InferenceSession("output/model.onnx")
      4 input_name = sess.get_inputs()[0].name
      5 print(input_name)

1 frames

/usr/local/lib/python3.7/dist-packages/onnxruntime/capi/onnxruntime_inference_collection.py in _create_inference_session(self, providers, provider_options, disabled_optimizers)
    356         session_options = self._sess_options if self._sess_options else C.get_default_session_options()
    357         if self._model_path:
--> 358             sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
    359         else:
    360             sess = C.InferenceSession(session_options, self._model_bytes, False, self._read_config_from_model)

Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from output/model.onnx failed:Fatal error: AliasWithName is not a registered function/op
```

I Thank you in advance!

**Comment by thiagocrepaldi** [comment]
I am working on adding support to ONNX export (with and without caffe2) https://github.com/facebookresearch/detectron2/pull/4120 and https://github.com/facebookresearch/detectron2/pull/4153/

Hopefully it gets accepted soon

**Comment by RajUpadhyay** [comment]
If you are still stuck on this, then you can use the **export_model.py** present in the tools/deploy/ folder of detectron2.
It works well since I've used it.

**Reply by bouachalazhar** [reply]
> If you are still stuck on this, then you can use the **export_model.py** present in the tools/deploy/ folder of detectron2. It works well since I've used it.

Can you give me an example ?

**Reply by RajUpadhyay** [reply]
Please refer [this](https://github.com/NVIDIA/TensorRT/tree/main/samples/python/detectron2) link

Only the first part if your goal is to just convert it into the onnx format.

for example, assuming you have satisfied the requirements of torch, torchvision, cuda and detectron2..

then run :

python detectron2/tools/deploy/export_model.py \
    --sample-image 1344x1344.jpg \
    --config-file detectron2/configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml \
    --export-method tracing \
    --format onnx \
    --output ./ \
    MODEL.WEIGHTS path/to/model_final_f10217.pkl \
    MODEL.DEVICE cuda
    
Please refer the above line link for clear instructions. 
Cheers!

**Reply by bouachalazhar** [reply]
> Please refer [this](https://github.com/NVIDIA/TensorRT/tree/main/samples/python/detectron2) link
> 
> Only the first part if your goal is to just convert it into the onnx format.
> 
> for example, assuming you have satisfied the requirements of torch, torchvision, cuda and detectron2..
> 
> then run :
> 
> python detectron2/tools/deploy/export_model.py --sample-image 1344x1344.jpg --config-file detectron2/configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml --export-method tracing --format onnx --output ./ MODEL.WEIGHTS path/to/model_final_f10217.pkl MODEL.DEVICE cuda
> 
> Please refer the above line link for clear instructions. Cheers!

I would like to convert maskrcnn to onnx model. For use export_model.py is it necessary to git clone detectron2 ?

**Reply by RajUpadhyay** [reply]
I am sorry but I do not quite get the question.
Have you not been using detectron2?

If not then yes, since this python file is in detectron2, please clone detectron2, and run setup.py

Follow the installation guide.

If I have misunderstood your question, do let me know.

**Reply by bouachalazhar** [reply]
> I am sorry but I do not quite get the question. Have you not been using detectron2?
> 
> If not then yes, since this python file is in detectron2, please clone detectron2, and run setup.py
> 
> Follow the installation guide.
> 
> If I have misunderstood your question, do let me know.

I have installed detectron2 : 0.6 with `python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'` and I use it but I don't see export_model.py

**Reply by RajUpadhyay** [reply]
Ah ok I see, then just clone it. Since you have already installed the detectron2 package which you can check via pip freeze | grep detectron2, just clone the git repo and you can use export_model.py.

**Reply by bouachalazhar** [reply]
> Ah ok I see, then just clone it. Since you have already installed the detectron2 package which you can check via pip freeze | grep detectron2, just clone the git repo and you can use export_model.py.

Ok I do it. Is it the best way for export detectron2 model to onnx ?

**Reply by thiagocrepaldi** [reply]
> > Ah ok I see, then just clone it. Since you have already installed the detectron2 package which you can check via pip freeze | grep detectron2, just clone the git repo and you can use export_model.py.
> 
> Ok I do it. Is it the best way for export detectron2 model to onnx ?

yes. another way is doing whatever the unit tests do at https://github.com/facebookresearch/detectron2/blob/main/tests/test_export_onnx.py

**Reply by bouachalazhar** [reply]
> > > Ah ok je vois, alors clonez-le. Puisque vous avez déjà installé le paquet detectron2 que vous pouvez vérifier via pip freeze | grep detectron2, il suffit de cloner le dépôt git et vous pouvez utiliser export_model.py.
> > 
> > 
> > Ok je le fais. Est-ce le meilleur moyen d’exporter le modèle detectron2 vers onnx ?
> 
> oui. Une autre façon est de faire ce que les tests unitaires font à https://github.com/facebookresearch/detectron2/blob/main/tests/test_export_onnx.py

How use it, can you give example ? If I understand there isn't function for export model when we install detectron2 package.

**Comment by 4daJKong** [comment]
Hello, have you solved your problem? I also face the same problem #4881

**Reply by bouachalazhar** [reply]
No yet

**Reply by RajUpadhyay** [reply]
Please refer this site.. 
https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch

Use the docker image to create the onnx and the corresponding tensorrt engine.

Note that the link I have mentioned is for x86 machine.

If you want to do it on jetson, please use l4t docker image.

**Reply by RajUpadhyay** [reply]
hi @bouachalazhar @4daJKong 
did you manage to create the onnx file successfully?
because due to some work, I had to generate onnx files again for faster_rcnn and I was able to generate it using **export_model.py**.

If not, let me know the model you are generating for and let me try in my setup.

**Reply by 4daJKong** [reply]
Thanks but not yet, I guess there is still some problem in my model, not in faster_rcnn...its backbone based on transformer-encoder, and convey some layers' feature to rcnn

**Reply by bouachalazhar** [reply]
> hi @bouachalazhar @4daJKong 
> 
> did you manage to create the onnx file successfully?
> 
> because due to some work, I had to generate onnx files again for faster_rcnn and I was able to generate it using **export_model.py**.
> 
> 
> 
> If not, let me know the model you are generating for and let me try in my setup.

I can't export maskrcnn and retinanet to onnx.

**Reply by RajUpadhyay** [reply]
@bouachalazhar I am OOTO but will I will get back to you tomorrow.
Tell me one thing, were you unable to generate onnx file for this?  config/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml
Because I generated one today itself so I will share with you and will let you know about my environment configs.
If you requirement is something else, let me know.

**Reply by bouachalazhar** [reply]
> @bouachalazhar I am OOTO but will I will get back to you tomorrow.
> 
> Tell me one thing, were you unable to generate onnx file for this?  config/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml
> 
> Because I generated one today itself so I will share with you and will let you know about my environment configs.
> 
> If you requirement is something else, let me know.
> 
> 

Yes I used export_model.py but I can't export, I think output or input shape it isn't good.

**Reply by RajUpadhyay** [reply]
@bouachalazhar Hi,
so I just generated a onnx file for mask_rcnn_R_50_FPN_3x.yaml.
I will share the command and my environment details and I will also share the onnx file with you.

**NOTE**: I still recommend using the docker image because I myself use docker because different version requirements can be painful.

**Environment:**
Ubuntu 20.04
Cuda: 12.0 (11.8 works to be honest, I upgraded my cuda for whole another project)
Cudnn: 8.6

onnx==1.13.1
onnxruntime==1.14.1
torch==1.13.1
torchaudio==0.13.1
torchvision==0.14.1

**Command**
python3 detectron2/tools/deploy/export_model.py --sample-image image.jpg --config-file detectron2/configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml  --export-method tracing --format onnx --output ./ MODEL.WEIGHTS .torch/iopath_cache/detectron2/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x/137849600/model_final_f10217.pkl

**ONNX file**
https://drive.google.com/file/d/1UzIHjDM4Qkq_KPh84G9AK6K7vuwqsK7L/view?usp=share_link

I hope this helps you in some way.

**Reply by bouachalazhar** [reply]
> @bouachalazhar Hi, so I just generated a onnx file for mask_rcnn_R_50_FPN_3x.yaml. I will share the command and my environment details and I will also share the onnx file with you.
> 
> **NOTE**: I still recommend using the docker image because I myself use docker because different version requirements can be painful.
> 
> **Environment:** Ubuntu 20.04 Cuda: 12.0 (11.8 works to be honest, I upgraded my cuda for whole another project) Cudnn: 8.6
> 
> onnx==1.13.1 onnxruntime==1.14.1 torch==1.13.1 torchaudio==0.13.1 torchvision==0.14.1
> 
> **Command** python3 detectron2/tools/deploy/export_model.py --sample-image image.jpg --config-file detectron2/configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml --export-method tracing --format onnx --output ./ MODEL.WEIGHTS .torch/iopath_cache/detectron2/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x/137849600/model_final_f10217.pkl
> 
> **ONNX file** https://drive.google.com/file/d/1UzIHjDM4Qkq_KPh84G9AK6K7vuwqsK7L/view?usp=share_link
> 
> I hope this helps you in some way.

It works for me, I think when I used git clone it does'nt work but my first problem it's to convert ONNX to OpenVINO IR(.xml file). Thx for all !

**Reply by RajUpadhyay** [reply]
Glad it is working.

**Reply by YingKn** [reply]
> @bouachalazhar Hi, so I just generated a onnx file for mask_rcnn_R_50_FPN_3x.yaml. I will share the command and my environment details and I will also share the onnx file with you.
> 
> **NOTE**: I still recommend using the docker image because I myself use docker because different version requirements can be painful.
> 
> **Environment:** Ubuntu 20.04 Cuda: 12.0 (11.8 works to be honest, I upgraded my cuda for whole another project) Cudnn: 8.6
> 
> onnx==1.13.1 onnxruntime==1.14.1 torch==1.13.1 torchaudio==0.13.1 torchvision==0.14.1
> 
> **Command** python3 detectron2/tools/deploy/export_model.py --sample-image image.jpg --config-file detectron2/configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml --export-method tracing --format onnx --output ./ MODEL.WEIGHTS .torch/iopath_cache/detectron2/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x/137849600/model_final_f10217.pkl
> 
> **ONNX file** https://drive.google.com/file/d/1UzIHjDM4Qkq_KPh84G9AK6K7vuwqsK7L/view?usp=share_link
> 
> I hope this helps you in some way.

Hi,RajU. I export onnx model by tracing too,but I can't load this model with Opencv function cv2::dnn::readNet.Have you tried to deploy your model in c++? Thank you in advance!

**Reply by RajUpadhyay** [reply]
@YingKn Unfortunately no, I haven't. I only needed the onnx file for the tensorrt engine.
But as far as I've seen, loading onnx model with cv2::dnn::readNetFromONNX() can be quite tricky. (for eg, dynamic input shapes dont work with opencv and all)
It would help to see the error but please see the opencv github issues, I have seen alot of issues related to dnn::read.

**Reply by YingKn** [reply]
> @YingKn Unfortunately no, I haven't. I only needed the onnx file for the tensorrt engine. But as far as I've seen, loading onnx model with cv2::dnn::readNetFromONNX() can be quite tricky. (for eg, dynamic input shapes dont work with opencv and all) It would help to see the error but please see the opencv github issues, I have seen alot of issues related to dnn::read.

ok RajU,thank you so much.
